server/depot.go

```diff
@@ -98,14 +112,93 @@ func (d fileDepot) Serial() (*big.Int, error) {
 		}
 		return s, nil
 	}
-	data, err := ioutil.ReadFile(name)
+	file, err := os.Open(name)
 	if err != nil {
 		return nil, err
 	}
-	serial := s.SetBytes(data)
+	defer file.Close()
+	r := bufio.NewReader(file)
+	// is there a better way??
+	data, err := r.ReadString('\r')
+	data = strings.TrimSuffix(data ,"\r")
+	data = strings.TrimSuffix(data ,"\n")
+	if err != nil && err != io.EOF {
+		return nil, err
+	}
+	serial, success := s.SetString(data,16)
+	if success != true  {
+		return nil, errors.New("Could not convert "+string(data)+" to serial number")
+	}
 	return serial, nil
 }
 
+func (d fileDepot) writeDB(cn string, serial *big.Int, filename string, cert *x509.Certificate) error {
+
+	var dn bytes.Buffer
+	var dbEntry bytes.Buffer
+
+	if err := os.MkdirAll(d.dirPath, 0755); err != nil {
+		return err
+	}
+	name := d.path("index.txt")
+
+	issuer := cert.Issuer
+
+	file, err := os.OpenFile(name, os.O_CREATE | os.O_RDWR | os.O_APPEND, dbPerm)
+	if err != nil {
+	    fmt.Printf("Could not append to "+name+" : %q\n", err.Error())
+	    return err
+	}
+	defer file.Close()
+
+	// Format of the caDB, see http://pki-tutorial.readthedocs.io/en/latest/cadb.html
+	//   STATUSFLAG  EXPIRATIONDATE  REVOCATIONDATE(or emtpy)    SERIAL_IN_HEX   CERTFILENAME_OR_'unknown'   Certificate_DN
+
+	serialHex  := fmt.Sprintf("%x", cert.SerialNumber)
+	t := cert.NotAfter
+	validDate := fmt.Sprintf("%d%02d%02dT%02d%02d%02dZ", t.Year(), t.Month(), t.Day(), t.Hour(), t.Minute(), t.Second())
+
+	if len(issuer.Country) >0 {
+	    dn.WriteString("/C=" + issuer.Country[0])
+	}
+	if len(issuer.Province) >0 {
+	    dn.WriteString("/ST=" + issuer.Province[0])
+	}
+	if len(issuer.Locality) >0 {
+	    dn.WriteString("/L=" + issuer.Locality[0])
+	}
+	if len(issuer.Organization) >0 {
+	    dn.WriteString("/O=" + issuer.Organization[0])
+	}
+	if len(issuer.OrganizationalUnit) >0 {
+	    dn.WriteString("/OU=" + issuer.OrganizationalUnit[0])
+	}
+	dn.WriteString("/CN="+cn)
+	if len(cert.EmailAddresses) >0 {
+	    dn.WriteString("/emailAddress=" + cert.EmailAddresses[0])
+	}
+	// Valid
+	dbEntry.WriteString("V\t")
+	// Valid till
+	dbEntry.WriteString(validDate+"\t")
+	// Emptry (not revoked)
+	dbEntry.WriteString("\t")
+	// Serial in Hex
+	dbEntry.WriteString(serialHex+"\t")
+	// Certificate file name
+	dbEntry.WriteString("\t"+filename+"\t")
+	// Certificate DN
+	dbEntry.Write(dn.Bytes());
+	dbEntry.WriteString("\n")
+
+	if _, err := file.Write(dbEntry.Bytes()); err != nil {
+		file.Close()
```

Comment:
by adding `defer file.Close()` above you're already ensuring that the file will be closed. 

I see that I made the same mistake below :)